File: 0147-insertion-sort-list/0147-insertion-sort-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def insertionSortList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode()
        dummy.next = head
        cur = head


        while cur and cur.next:
            if cur.val <= cur.next.val:
                cur = cur.next
            else:
                lesser_element = dummy
                while lesser_element.next.val < cur.next.val:
                    lesser_element = lesser_element.next

                temp = cur.next
                cur.next = temp.next
                temp.next = lesser_element.next
                lesser_element.next = temp

        return dummy.next

File: 0147-insertion-sort-list/0147-insertion-sort-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
public class Solution {
    public ListNode insertionSortList(ListNode head) {
        ListNode dummy = new ListNode();
        dummy.next = head;
        ListNode cur = head;

        while (cur != null && cur.next != null) {
            if (cur.val <= cur.next.val) {
                cur = cur.next;
            } else {
                ListNode lesserElement = dummy;
                while (lesserElement.next != null && lesserElement.next.val < cur.next.val) {
                    lesserElement = lesserElement.next;
                }

                ListNode temp = cur.next;
                cur.next = temp.next;
                temp.next = lesserElement.next;
                lesserElement.next = temp;
            }
        }

        return dummy.next;
    }
}